### PR TITLE
cmake: fix link errors on x86_64 Linux

### PR DESCRIPTION
**Pull request by davvid**
libbbl fails to link on x86_64 Linux with the following error:
    
        /usr/bin/ld: bbl/libbbl.a(api_capi.cpp.o):
        relocation R_X86_64_PC32 against symbol '_ZTISt12out_of_range@@GLIBCXX_3.4'
        can not be used when making a shared object; recompile with -fPIC
    
Enable position-independent code in cmake so that -fPIC is used when compiling.

Remove a reference to `process.hpp` which seems to be missing and unused.

Add missing `cstdio` includes.

### CODE REVIEW

File: bbl/src/cpp_context.cpp [hunk 1]
```diff
@@ -2,7 +2,6 @@
 #include "astutil.hpp"
 #include "bbl-detail.h"
 #include "bblfmt.hpp"
-#include "process.hpp"
```
Comment (davvid) [inline bbl/src/cpp_context.cpp:5]: Let me know if I'm missing something related to `process.hpp`.

Is this file perhaps sitting uncommitted in your tree? Things seem to compile fine without it so I removed this `#include` but I'm on Linux so my build environment might be different.

**Comment by anderslanglands** [comment]
Thanks for the PR! I've just pushed my current commits which remove that
#include. The other issues you're seeing are interesting as I'm building on
both Windows and WSL2 and not getting the same... In any case could you
rebase your changes on head please?

In general if you're just looking to try it out, you'd be better off
checking out the latest release tag - I'm pushing constantly back and forth
between Windows and WSL on head so it will often be broken in one and not
the other, even though I try and keep it working.

**Comment by davvid** [comment]
All good, I don't mind some breakage. I prefer building from source so that I can help improve portability-related edge cases.

I'm on the latest Debian/testing (rolling release) using llvm/clang 16, gcc 13.2.0 and glibc 2.37. The Debian packages I installed were `llvm-dev` and `libclang-dev`. I'll keep rebasing as new commits show up.

I pushed up a few more commits that got the test suite passing here. I split them up into small commits so feel free to cherry-pick a subset of the `build` branch if there's some you'd prefer over the others.

The `run_test.py` modification I had to make to get things passing is interesting. I'm curious whether the omission of `-Ibuild/include` works in WSL + Windows as well. I suspect that and the `cstdint`-related changes could be related to the newer gcc / glibc versions over here. I included the errors in the commit messages in case they're helpful. If that ends up being different over there I don't mind turning that change into a command-line option for `run_tests.py`.

Quick copy/paste snippet to make these commits available in your worktree for review:
```
git remote add davvid https://github.com/davvid/babble.git
git fetch davvid
git log -p --decorate davvid/build HEAD
```

**Comment by anderslanglands** [comment]
Thanks for kicking the tyres I appreciate it!

The -Ibuild/include is interesting as I have to specifically add that to get it to find the headers in WSL. I wonder, since you installed LLVM from the system package manager is it actually using GCC to build or is it using clang?

**Comment by anderslanglands** [comment]
Yep your branch works fine on Windows but the tests fail to run on WSL because it can't find stddef.h without that include. Hmmm will need to think about how to fix this...

**Comment by anderslanglands** [comment]
OK I'm going to merge then detect WSL in the test runner and add the include in that case.

**Comment by davvid** [comment]
> I wonder, since you installed LLVM from the system package manager is it actually using GCC to build or is it using clang?

Yes, `bbl-translate` was being built with gcc. I did try building it with `clang++` as well but that ended up with the same result so I didn't pursue that much further. Thanks for adding the platform detection.

I believe the root cause was the header inclusion order. Over on Linux `/usr/include` exists and the system headers in `build/include` shadow them in a way that makes the `#include_next <inttypes.h>` in `build/include/inttypes.h` eventually include clang's build/include/stdint.h rather than the system one needed by other system headers.

If my diagnosis is correct then unconditionally using `-idirafter build/include` (on all platforms) would resolve the issue since it'll keep any pre-existing paths at the front.

Does `-idirafter build/include` work on WSL? If so that might be more amenable to different setups instead of the platform detection.

[This page](https://clangd.llvm.org/guides/system-headers#how-clangd-finds-those-headers)  documents the interactions between clang and system headers. My reading of it does suggest that the path order is the root cause.

Using the debugging tip on that page, here are the paths on this system (with a bunch of other details omitted for brevity). Here's the default search path without any flags:

```bash
$ clang -v -c -xc++ /dev/null
#include <...> search starts here:
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/backward
 /usr/lib/llvm-16/lib/clang/16/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

Here's what we get when we add `-idirafter build/include`. Note how `build/include` ends up at the end.

```
$ clang -v -c -xc++ /dev/null -idirafter build/include
#include <...> search starts here:
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/backward
 /usr/lib/llvm-16/lib/clang/16/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
 build/include
End of search list.
```

Lastly, here's `-Ibuild/include` which runs into errors over here. Note how `build/include` is listed before any of the system paths. I think this explains it.

```
$ clang -v -c -xc++ /dev/null -Ibuild/include
#include <...> search starts here:
 build/include
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/x86_64-linux-gnu/c++/13
 /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/backward
 /usr/lib/llvm-16/lib/clang/16/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

Thanks Anders, I hope that helps.

**Comment by anderslanglands** [comment]
Does indeed seem to work on both Windows and WSL. Thanks!

**Pull request merged**